rfcs/0002-borrow-type.md

```diff
@@ -0,0 +1,277 @@
+
+It might seem we're out of luck.
+
+##  Struct Fields
+
+Sometimes the Windows APIs will declare a struct that has a non-"blittable" type as a field. These fields are logically "borrowed". For example, if a struct had an `HSTRING` as a field, it would not be responsibility of whoever constructs the struct to first increment the reference count. Also, when the struct is no longer needed, it is not necessary to decrement the reference count. 
+
+The struct as a whole implicitly has a lifetime associated with it whereby the fields of the struct must live for at least as long as the struct as a whole.
+
+Like "IN" params, using a reference `&T` is not possible since the field must be exactly equivalent to `T` in-memory. And `ManuallyDrop` also has the problem of consuming the `T` and requiring when the `T` is no longer used, the field is in scope so that `ManuallyDrop::into_inner` can be called.
+
+# Guide-level Explanation  
+[guide-level-explanation]: #guide-level-explanation
+
+***Note**: The following is what documentation for this feature might look like.*
+
+You may see the type `Borrow<'a, T>` used in Windows APIs. There's no need to worry about this type, it's actually quite straight forward.
+
+`Borrow<'a, T>` is exactly equivalent to `&'a T` except that in memory a `Borrow<'a, T>` looks just like a `T`.
+
+This is important because some Windows APIs take arguments of some type `T` but treat them as they are just borrowed.
+
+Let's take a look at an example:
+
+`windows::Data::Json::JsonObject` has a function named `Parse` which looks like this:
+
+```rust 
+fn Parse<'a>(input: Borrow<'a, HSTRING>) -> Result<JsonObject>;
+```
+
+The reason `Parse` takes a `Borrow<'a, HSTRING>` instead of just a plain `HSTRING` is because if it did, ownership of `input` would be passed into `Parse` and `input`'s `Drop` implementation would be called possibly freeing `input`'s memory!
+
+`Parse` really expects to logically borrow `input`. If this were a normal Rust function, you would expect `input` to be something like `&HSTRING`. However, `Parse` is not a regular Rust function - it crosses the FFI boundary into another language entirely! If were were to try to pass a `&HSTRING` into `Parse`, `Parse` would interpret the reference as if it were just a plain old `HSTRING`. This would cause all kinds of problem's all caused by the fact that `Parse` expected an `HSTRING` and we tried to give it a `&HSTRING`.
+
+`Borrow<'a, T>` allows us to emulate a reference but at the memory level still pass around memory that looks just like the underlying `T`. We get the safety of a Rust reference when writing our Rust code, and the Windows API receives the memory it expects to. Everyone wins!
+
+# Reference-level Explanation  
+[reference-level-explanation]: #reference-level-explanation
+
+`Borrow<'a, T>` has the following definition:
+
+```rust
+#![warn(unsafe_op_in_unsafe_fn)]
+
+#[repr(transparent)]
+struct Borrow<'a, T> {
```

Comment:
I agree with @nrc, for consistency it should probably be called `Borrowed`, primarily used for fields with a `In` type alias used by code gen for input parameters.

Reply:
I'm fine with the `Borrowed` nomenclature. I'm unsure if providing an `In` alias would be helpful. While it would make generated code shorter and thus likely a bit easier to parse, it would require the reader to understand that `In` and `Borrowed` are the same thing.

Reply:
Sure, we can always add it later if it becomes obviously valuable.